src/Microsoft.AspNetCore.Mvc.Core/Formatters/OutputFormatter.cs

```diff
@@ -112,17 +129,36 @@ public virtual bool CanWriteResult(OutputFormatterCanWriteContext context)
             }
             else
             {
-                // Confirm this formatter supports a more specific media type than requested e.g. OK if "text/*"
-                // requested and formatter supports "text/plain". contentType is typically what we got in an Accept
-                // header.
+                // Confirm this formatter 
                 var parsedContentType = new MediaType(context.ContentType);
                 for (var i = 0; i < SupportedMediaTypes.Count; i++)
                 {
                     var supportedMediaType = new MediaType(SupportedMediaTypes[i]);
-                    if (supportedMediaType.IsSubsetOf(parsedContentType))
+                    if (supportedMediaType.HasWildcard)
+                    {
+                        // For supported media types that are wildcard patterns, confirm that the requested
+                        // media type satisfies the wildcard pattern (e.g., if "text/entity+json;v=2" requested
+                        // and formatter supports "text/*+json").
+                        // We only do this when comparing against server-defined content types (e.g., those
+                        // from [Produces] or Response.ContentType), otherwise we'd potentially be reflecting
+                        // back arbitrary Accept header values.
```

Comment:
I think that this makes intuitive sense. The extra property is a little bit gross, but I think we'll have to live with it.